types/vimeo/index.d.ts

```diff
@@ -1,30 +1,166 @@
-// Type definitions for Vimeo
-// Project: http://developer.vimeo.com/player/js-api
-// Definitions by: Daz Wilkin <https://github.com/DazWilkin>
+// Type definitions for Vimeo 2.1
+// Project: https://github.com/vimeo/vimeo.js
+// Definitions by: Matthew Leffler <https://github.com/mattleff>
 // Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
 
-interface VimeoParams {
-	name:string;
-	value:any;
-}
+// tslint:disable-next-line
+declare module 'vimeo' {
```

Comment:
Rather than using `declare module`, I recommend just following [this template](https://www.typescriptlang.org/docs/handbook/declaration-files/templates/module-d-ts.html) and directly exporting the contained members.

Reply:
I've tried that but I get the following error at a call site (i.e., `import ... from 'vimeo';`):

> Could not find a declaration file for module 'vimeo'. '/Users/matthew/Development/alc-backend/node_modules/vimeo/index.js' implicitly has an 'any' type.
> Try `npm install @types/vimeo` if it exists or add a new declaration (.d.ts) file containing `declare module 'vimeo';`ts(7016)

What am I missing here?

Reply:
Can you describe your project structure when you take that approach? 
That is, where is this `index.d.ts` file in relation to the file containing the import statement?

Reply:
Thanks for your reply! Here's roughly how things look:

```
/server/common/vimeo.ts // file with import statement
/types/vimeo/index.d.ts // PRs file
/node_modules/vimeo/index.js // the JS module
```

Here's my `tsconfig.json`:

```json
{
	"compileOnSave": false,
	"compilerOptions": {
		"baseUrl": ".",
		"paths": {
			"@app/*": ["server/*"],
			"@controllers/*": ["server/api/controllers/*"],
			"@models/*": ["server/api/models/*"],
			"@services/*": ["server/api/services/*"],
			"@common/*": ["server/common/*"]
		},
		"target": "es2016",
		"module": "commonjs",
		"sourceMap": true,
		"strict": true,
		"noUnusedLocals": true,
		"typeRoots": ["node_modules/@types", "./types"]
	},
	"include": ["server/**/*.ts", "spec/**/*.ts", "scripts/**/*.ts"],
	"exclude": ["node_modules"]
}
```

Reply:
Interesting, that looks like that should work... I presume there are other @types packages contained in that `types` folder? If so, do others follow the same structure I recommended?

Reply:
All the `@types/*` packages are in `node_modules/@types/...`. `./types/...` is for types I have written or extensions on other `@types/*` types.  I know that TypeScript is picking up the custom types in `./types` because if I remove that from my `tsconfig.json` it will complain about the missing definitions I've added there. If I remove the `declare module '...'` for any of the types in `./types` I get errors.

Reply:
I finally found this: https://github.com/microsoft/TypeScript/issues/22217#issuecomment-369783776

> `"typeRoots"` is meant for global code. i.e. something that is declarated in the global namespace, and you want to include it. this is why your `declare module 'graphql-list-fields' {..` works, since it just declares a module with that name in the global namespace.

So `typeRoots` doesn't work like I thought. [This](https://stackoverflow.com/a/40326719/168581) helped me understand how the types are loaded which led me to the following `tsconfig.json`:

```json
{
	"compileOnSave": false,
	"compilerOptions": {
		"baseUrl": ".",
		"paths": {
			"@app/*": ["server/*"],
			"@controllers/*": ["server/api/controllers/*"],
			"@models/*": ["server/api/models/*"],
			"@services/*": ["server/api/services/*"],
			"@common/*": ["server/common/*"],
			"*": ["./types/*"]
		},
		"target": "es2016",
		"module": "commonjs",
		"sourceMap": true,
		"strict": true,
		"noUnusedLocals": true
	},
	"include": ["server/**/*.ts", "spec/**/*.ts", "scripts/**/*.ts", "types/**/*.ts"]
}
```

One that was in place directly exporting contained members worked fine. Try out 024d949.

Reply:
Wow, I had no idea... now we know, I guess.